src/contract.rs

```diff
@@ -53,127 +58,160 @@ pub fn execute(
+            deps, timestamp, from, to,
+        )?),
+        QueryMsg::GetValidatorMetricsBtwTimestamps {
+            addr,
+            timestamp1,
+            timestamp2,
+        } => to_binary(&query_all_validator_metrics_btw_timestamps(
+            deps, addr, timestamp1, timestamp2,
+        )?),
     }
 }
 
+fn query_timestamps(deps: Deps) -> StdResult<Vec<u64>> {
+    Ok(STATE.load(deps.storage)?.cron_timestamps)
+}
+
 fn query_validator_apr(
     deps: Deps,
     timestamp1: u64,
     timestamp2: u64,
     addr: Addr,
 ) -> StdResult<ValidatorAprResponse> {
-    if timestamp1.eq(&timestamp2) {
+    if timestamp1.ge(&timestamp2) {
         return Err(StdError::GenericErr {
-            msg: "timestamp1 and timestamp2 cannot be the same".to_string(),
+            msg: "timestamp1 cannot be greater than or equal to timestamp2".to_string(),
         });
     }
 
-    let h1_opt = METRICS_HISTORY
-        .load(deps.storage, U64Key::new(timestamp1))?
-        .into_iter()
-        .find(|history| history.addr.eq(&addr));
+    let h1 = METRICS_HISTORY.load(deps.storage, (&addr, U64Key::new(timestamp1)))?;
 
-    let h2_opt = METRICS_HISTORY
-        .load(deps.storage, U64Key::new(timestamp2))?
-        .into_iter()
-        .find(|history| history.addr.eq(&addr));
-
-    if h1_opt.is_none() || h2_opt.is_none() {
-        return Err(StdError::GenericErr {
-            msg: "Validator does'nt have metrics recorded for all the given times".to_string(),
-        });
-    }
+    let h2 = METRICS_HISTORY.load(deps.storage, (&addr, U64Key::new(timestamp2)))?;
 
     return Ok(ValidatorAprResponse {
         addr,
-        apr: compute_apr(&h1_opt.unwrap(), &h2_opt.unwrap(), timestamp2 - timestamp1),
+        apr: compute_apr(&h1, &h2, timestamp2 - timestamp1),
     });
 }
 
-fn convert_validator_metrics_to_map(
-    metrics: Vec<ValidatorMetrics>,
-) -> HashMap<Addr, ValidatorMetrics> {
-    let mut map: HashMap<Addr, ValidatorMetrics> = HashMap::new();
-    for metric in metrics {
-        map.insert(metric.addr.clone(), metric);
-    }
-    map
-}
-
-fn query_all_validators_aprs(
+fn query_validators_aprs_by_interval(
     deps: Deps,
     timestamp1: u64,
     timestamp2: u64,
+    from: u32,
+    to: u32,
 ) -> StdResult<Vec<ValidatorAprResponse>> {
-    if timestamp1.eq(&timestamp2) {
+    if timestamp1.ge(&timestamp2) {
         return Err(StdError::GenericErr {
-            msg: "timestamp1 and timestamp2 cannot be the same".to_string(),
+            msg: "timestamp1 cannot be greater than or equal to timestamp2".to_string(),
         });
     }
+    let validators = STATE.load(deps.storage)?.validators;
 
-    let mut response: Vec<ValidatorAprResponse> = vec![];
+    let total_validators: u32 = validators.len().try_into().unwrap();
 
-    let history1_map = convert_validator_metrics_to_map(
-        METRICS_HISTORY.load(deps.storage, U64Key::new(timestamp1))?,
-    );
+    if to.ge(&total_validators) || from > to {
```

Comment:
Is the expectation that manager will know how many validators there are in KYV at all times?

Reply:
We can expose a query for no of validators, right? I feel like building this way makes the contract simpler and less error-prone in itself - when we can make these types of computations in the frontend itself where rewriting/revisiting things are allowed.